File: PRACTICAS/PRACTICA_ACTUAL/RUBY/lib/titulo_propiedad.rb
# encoding: utf-8

module Civitas
class TituloPropiedad
  @@factor_intereses_hipoteca=1.1
  
  
  def initialize(nom,ab,fr,hb,pc,pe)
    @nombre=nom
    @alquiler_base=ab
    @factor_revalorizacion=fr
    @hipoteca_base=hb
    @precio_compra=pc
    @precio_edificar=pe
    @num_casas=0
    @num_hoteles=0
    @hipotecado=false
    @propietario = nil
  end
  
  attr_reader   :hipotecado,  :nombre, :num_casas, :num_hoteles, :precio_compra, :precio_edificar, :propietario
  
#  ATTR get_hipotecado, get_importe_hipoteca, get_importe_cancelar_hipoteca....get_propietario
  def get_importe_hipoteca
    @hipoteca_base*(1+@num_casas*0.5+@num_hoteles*2.5)
  end
  
  def get_importe_cancelar_hipoteca
    (@hipoteca_base*(1+@num_casas*0.5+@num_hoteles*2.5))*@@factor_intereses_hipoteca
  end
  
  def get_precio_alquiler
    if @hipotecado||propietario_encarcelado
      return 0
    else
      return @alquiler_base*(1+(@num_casas*0.5)+(@num_hoteles*2.5))
    end
  end
  
  def get_precio_venta
    @precio_compra+(@num_casas+5*@num_hoteles)*@precio_edificar*@factor_revalorizacion
  end
  
  def actualiza_propietario_por_conversion(jugador)
    @propietario=jugador
  end
  
  def cancelar_hipoteca(jugador)
    operacion=false
    if @hipotecado&&jugador==@propietario
      operacion=true
      @propietario.paga(get_importe_cancelar_hipoteca)
      @hipotecado=false
    end
      operacion
  end
  
  def cantidad_casas_hoteles
    @num_casas+@num_hoteles
  end
  
  def comprar(jugador)
    operacion=false
    if @propietario==nil
      @propietario=jugador
      @propietario.paga(@precio_compra)
      operacion=true
    end
    return operacion
  end
  
  def construir_casa(jugador)
    operacion=false
    if es_este_el_propietario(jugador)
      @propietario.paga(@precio_edificar)
      @num_casas+=1
      operacion=true
    end
    return operacion
  end
  
  def construir_hotel(jugador)
    operacion=false
    if es_este_el_propietario(jugador)
      operacion=true
      @propietario.paga(@precio_edificar)
      @num_casas-=4
      @num_hoteles+=1
    end
    return operacion
  end
  
  def derruir_casas(n,jugador)
    operacion=false
    if es_este_el_propietario(jugador)&&@num_casas>=n
      operacion=true
      @num_casas-=n
    end
    return operacion
  end
  
  def es_este_el_propietario(jugador)
    return @propietario==jugador
  end
  
  def hipotecar(jugador)
    operacion=false
    if !@hipotecado&&es_este_el_propietario(jugador)
      operacion=true
      @propietario.recibe(@hipoteca_base*(1+(@num_casas*0.5)+(@num_hoteles*2.5)))
      @hipotecado=true
    end
    return operacion
  end
  
  def propietario_encarcelado
    if tiene_propietario
      return @propietario.is_encarcelado
    else
      return false
    end
  end
  
  def tiene_propietario
    return @propietario!=nil
  end
  
  def to_s
    return      "\n  alquilerBase: "+@alquiler_base.to_s+
                "\n  factorInteresesHipoteca: "+@@factor_intereses_hipoteca.to_s+
                "\n  factorRevalorizacion: "+@factor_revalorizacion.to_s+
                "\n  hipotecaBase:"+@hipoteca_base.to_s+
                "\n  hipotecado: "+@hipotecado.to_s+
                "\n  numCasas:"+@num_casas.to_s+
                "\n  numHoteles:"+@num_hoteles.to_s+
                "\n  precioCompra: "+@precio_compra.to_s+
                "\n  precioEdificar:"+@precio_edificar.to_s+
                "\n  propietario: "+@propietario.to_s;
  end
  
  def tramitar_alquiler(jugador)
    if tiene_propietario&&!es_este_el_propietario(jugador)
      cantidad=@alquiler_base*(1+(@num_casas*0.5)+(@num_hoteles*2.5))
      jugador.paga_alquiler(cantidad)
      @propietario.recibe(cantidad)
    end
  end
  
  def vender(jugador)
    operacion=false
    if es_este_el_propietario(jugador)&&!@hipotecado
      operacion=true
      @propietario.recibe(get_precio_venta)
      @propietario=nil
      @num_casas=0
      @num_hoteles=0
    end
    return operacion
  end
end
end

File: PRACTICAS/practica1/Civitas_P1/src/civitas_p1/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

/**
 *
 * @author xaviv
 */
public class TituloPropiedad {
    
    private String nombre;
    private float alquilerBase;
    private float factorRevalorizacion;
    private final float factorInteresHipoteca = (float) 1.1;
    private float hipotecaBase;
    private float PrecioCompra;
    private float PrecioEdificar;
    private boolean hipotecado;
    private int numCasas;
    private int numHoteles;
    private Jugador propietario;
    
    TituloPropiedad(String nom,float base_alq,float factor_rev,float base_hipot,float precio_comp,float precio_edif){
        
        nombre=nom;
        alquilerBase=base_alq;
        factorRevalorizacion=factor_rev;
        hipotecaBase=base_hipot;
        PrecioCompra=precio_comp;
        PrecioEdificar=precio_edif;
        propietario = null;
        numCasas = 0;
        numHoteles = 0;
        hipotecado=false;
    }
    
    
    public String toString(){
        
        if(hipotecado)
            return "La propiedad " + nombre + " es de " + propietario.getNombre() + " con un alquiler base de " + alquilerBase +  ", una hipoteca base de " + hipotecaBase + ", un precio de compra de " + PrecioCompra + ", un precio de edificacion de " + PrecioEdificar + " " +numCasas+ " casas y " +numHoteles + " hoteles";
        else
            return "La propiedad " + nombre + " no está hipotecada por nadie. Tiene un alquiler base de " + alquilerBase +", una hipoteca base de " +hipotecaBase+ ", un precio de compra de " +PrecioCompra +", un precio de edificacion de " +PrecioEdificar +", " +numCasas +" casas y " +numHoteles + " hoteles" ;
            
    }
    private float getPrecioAlquiler(){
        if(hipotecado || propietario.propietarioEncarcelado())
            return 0;
        else
            return (float) (alquilerBase*(1+(numCasas*0.5)+(numHoteles*2.5)));
        
    }
    float getImporteCancelarHipoteca(){
        
        return hipotecaBase*factorInteresHipoteca;
        
    }
    boolean cancelarHipoteca(Jugador jugador){
        
        boolean sehace=false;
        
        if(hipotecado && esEsteElPropietario(jugador)){
            
            propietario.paga(getImporteCancelarHipoteca());
            hipotecado=false;
            sehace =true;
            
        }
        
        return sehace;
        
    }
    boolean hipotecar(Jugador jugador){
        
        boolean sehace = false;
        
        if(!hipotecado && esEsteElPropietario(jugador)){
            propietario.recibe(hipotecaBase);
            hipotecado=true;
            sehace = true;
        }
        
        return sehace;
        
    }
    void tramitarAlquiler(Jugador jugador){
        
        if(propietario != null && !esEsteElPropietario(jugador)){
            
            jugador.pagaAlquiler(getPrecioAlquiler());
            propietario.recibe(getPrecioAlquiler());
            
        }
        
    }
    private boolean propietarioEncarcelado(){
        
        if(!propietario.propietarioEncarcelado() || propietario == null)
            return false;
        else
            return true;
        
    }
    int cantidadCasasHoteles(){
        
        return numCasas+numHoteles;
        
    }
    boolean derruirCasas(int n, Jugador jugador){
        
        boolean sehace=false;
        
        if(numCasas>=n && esEsteElPropietario(jugador)){
            
            numCasas=numCasas-n;
            sehace = true;
            
        }
        
        return sehace;
        
    }
    private float getPrecioVenta(){
        
        return PrecioCompra+cantidadCasasHoteles()*PrecioEdificar*factorRevalorizacion;
        
    }
    boolean construirCasa(Jugador jugador){
        
        boolean sehace=false;
        
        if(esEsteElPropietario(jugador)){
            
            jugador.paga(PrecioEdificar);
            numCasas++;
            sehace=true;
            
        }
        
        return sehace;
        
    }
    boolean construirHotel(Jugador jugador){
        
        boolean sehace=false;
        
        if(esEsteElPropietario(jugador)){
            
            jugador.paga(PrecioEdificar);
            numHoteles++;
            sehace=true;
            
        }
        
        return sehace;
        
    }
    boolean Comprar(Jugador jugador){
        
        if(propietario!=null){
            
            propietario=jugador;
            jugador.paga(PrecioCompra);
            return true;
            
        }else{
            
            return false;
            
        }
        
    }
    boolean tiene_propietario(){
        
        return propietario != null;
        
    }
    
    void actualizaPopietarioPorConversion(Jugador jugador){
        
        propietario=jugador;
        
    }
    private boolean esEsteElPropietario(Jugador jugador){
        
        return propietario==jugador;
        
    }        
    boolean vender(Jugador jugador){
        
        if(jugador.getSaldo()>= PrecioCompra &&jugador!=propietario){
            jugador.paga(PrecioCompra);
            propietario=jugador;
            return true;
        }else{
            return false;
        }
    }
    private float getImporteHipoteca(){
        
        return (float) (hipotecaBase*(1+(numCasas*0.5)+(numHoteles*2.5)));

        
    }
    int getNumCasas(){
        
        return numCasas;
        
    }
    int getNumHoteles(){
        
        return numHoteles;
        
    }
    float getPrecioCompra(){
        
        return PrecioCompra;
        
    }
    float getPrecioEdificar(){
        
        return PrecioEdificar;
        
    }
    Jugador getPropietario(){
        
        return propietario;
        
    }
    
       
}
